lua/entities/gmod_tardis/modules/sh_health.lua

```diff
@@ -151,16 +151,96 @@ if SERVER then
 		self:CallHook("RepairStarted")
 	end
 
+	local migrated_data_names = {
+		"floatfirst",
+		"hads",
+		"spindir",
+		"spindir_prev",
+		"spindir_before_door",
+		"handbrake",
+		"physlock",
+		"demat-fast",
+		"fastreturn-pos",
+		"fastreturn-ang"
+	}
+
 	function ENT:FinishRepair()
 		if self:CallHook("ShouldRedecorate") then
-			local ent = TARDIS:SpawnTARDIS(self:GetCreator(),{
-				metadataID = TARDIS:GetSetting("redecorate-interior","default",self:GetCreator()),
+			local pos = self:GetPos() + Vector(0,0,0.02)
+			local ang = self:GetAngles()
+
+			local ent = TARDIS:SpawnTARDIS(self:GetCreator(), {
+				metadataID = self:GetData("redecorate-interior"),
 				finishrepair = true,
-				pos = self:GetPos()+Vector(0,0,2),
-				ang = self:GetAngles()
+				pos = pos,
+				ang = ang,
+				parent = self,
 			})
-			self:Remove()
+			ent:SetBodygroup(1,0)
+
+			-- let them know each other
+			self:SetData("redecorate_next", ent)
+
+			-- make it dematerialised
+			ent:SetData("demat",true)
+			ent:SetData("vortex", 1)
+			ent:SetData("vortexalpha", 1)
+			ent:SetData("teleport",true)
+			ent:SetData("alpha", 0)
+			ent:DrawShadow(false)
+			for k,v in pairs(ent.parts) do
+				v:DrawShadow(false)
+			end
+
+			ent:StopDemat()
+			ent:SetCollisionGroup(COLLISION_GROUP_IN_VEHICLE)
+
+			-- position
+			ent:SetPos(pos)
+			ent:SetAngles(ang)
+			ent:SetData("demat-pos",pos,true)
+			ent:SetData("demat-ang",ang,true)
+			ent:SetDestination(pos, ang)
+
 			ent:GetPhysicsObject():Sleep()
+
+			-- make it alive again
+			self:SetData("repairing", false, true)
+			self:ChangeHealth(TARDIS:GetSetting("health-max"))
+			self:SetPower(true)
+
+
+			-- migrate data
+			local saved_data = {}
+
+			for k,v in ipairs(migrated_data_names) do
+				saved_data[v] = self:GetData(v)
+			end
+
+			ent:SetData("cloak", self:GetData("cloak"), true)
+
+			ent:SetData("parent-saved-data", saved_data, true)
+			ent:SetData("parent-security", self.interior:GetData("security", false), true)
+			ent:SetData("prevortex-flight", self:GetData("flight", false), true)
+
+			-- sonic (if exists)
+			local owner = self:GetCreator()
+			if owner and owner.linked_tardis == self then
+				owner.linked_tardis = ent
+				net.Start("Sonic-SetLinkedTARDIS")
+					net.WriteEntity(ent)
+				net.Send(owner)
+			end
+
+
+			-- fly away
+			self:SetData("redecorate-demattime", CurTime())
+			self:SetFastRemat(true)
+			self:SetPhyslock(true) -- should not move
+			ent:SetPhyslock(true)
```

Comment:
This results in a weird experience, is this strictly necessary especially with the parenting stuff added in now?